### PR TITLE
Add completion support for regular expressions

### PR DESCRIPTION
**Pull request by TharushiJay**
## Purpose
This PR adds completion support for regular expressions with 
1. the regex snippet `re` `` 
2. interpolation (already supported)
3. snippets for character classes `d, D, s, S, w, W` and control escape `r, n, t` .
4. snippets for `()` and `[]`
5. snippets for flag options `m,s,i,x`

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/38095

## Samples
1.
Screenshot 2022-11-01 at 13 22 03

3.
Screenshot 2022-11-01 at 13 22 21

4.
Screenshot 2022-11-04 at 09 57 43

5.
Screenshot 2022-11-04 at 10 23 50

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/TemplateExpressionNodeContext.java [hunk 1]
```diff
@@ -81,6 +90,36 @@ public List<LSCompletionItem> getCompletions(BallerinaCompletionContext context,
...
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_WHITESPACE.get()));
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_NON_WHITESPACE.get()));
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_ALPHA_NUMERIC.get()));
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_NON_ALPHA_NUMERIC.get()));
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_RETURN.get()));
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_NEWLINE.get()));
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_TAB.get()));
+        } else if (nodeAtCursor.kind() == SyntaxKind.RE_FLAG_EXPR 
+                && ((ReFlagExpressionNode) nodeAtCursor).questionMark().position() + 1 == ctx
+                .getCursorPositionInTree()) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/TemplateExpressionNodeContext.java:109]: Shall we move this condition to a separate method with a proper name?

**Review by malinthar — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/TemplateExpressionNodeContext.java (hunk 1 above)
Comment (TharushiJay) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/TemplateExpressionNodeContext.java]: Done

**Review by malinthar — APPROVED** [review]
LGTM!

**Comment by pcnfernando** [comment]
@TharushiJay added a comment in https://github.com/ballerina-platform/ballerina-lang/issues/38095#issuecomment-1309743305. Is that do-able?

**Comment by mohanvive** [comment]
> #38095 (comment)

@pcnfernando I wanted to understand this requirement a bit more. Let's have a chat on this later.

**Review by IMS94 — APPROVED** [review]
LGTM!

**Pull request merged**